lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/NodeStruct.cs

```diff
@@ -62,10 +67,29 @@ internal NodeStruct(string id, string groupName, string operatorName)
         /// <summary>
         /// Gets the first message in the message queue.
         /// </summary>
+        /// <param name="cancellationSource">The cancellation token for the data reading operation cancellation</param>
         /// <returns>The first available message.</returns>
-        internal T[] GetData()
+        internal T[] GetData(CancellationTokenSource cancellationSource = null)
         {
-            return _messageQueue.Take().Data;
+            try
+            {
+                if (cancellationSource == null || !cancellationSource.IsCancellationRequested)
+                {
+                    var r = cancellationSource == null
+                        ? _messageQueue.Take().Data
+                        : _messageQueue.Take(cancellationSource.Token).Data;
+                    return r;
+                }
+                else
+                {
+                    throw new OperationCanceledException("GetData operation is canceled");
+                }
+            }
+            catch (OperationCanceledException e)
+            {
+                Logger.Log(Level.Warning, "Received OperationCanceledException in NodeStruct.GetData() with message {0}.", e.Message);
+                throw e;
```

Comment:
That's true. Just for debug purpose, I am watching where it is thrown originally and  where it is caught finally in the task.